### PR TITLE
[AMQ-9107] Remove consumers more efficiently in ManagedRegionBroker

### PR DESCRIPTION
**Pull request by lucastetreault**
Running a profiler while executing the sample code attached to [AMQ-9107](https://issues.apache.org/jira/browse/AMQ-9107) identified ManagedRegionBroker.removeConsumer as the bottleneck. The existing implementation loops over all the subscriptions to find the subscription for the consumer we want to close. When we have n consumers and we want to close them all this for loop is O(n^2) and when n is big enough it creates a serious performance issue. With 188,000 consumers we observe the CPU at 100% for ~40 minutes while all the connections are closed: 

image

After this PR, running the same test case we observe a spike in CPU of only one minute or less, similar to what it took to create the consumers: 

image

I ran the full suite of tests and everything is passing.

### CODE REVIEW

**Review by jbonofre — APPROVED** [review]

**Comment by mattrpav** [comment]
In the doStop()-- shouldn't all the cache collections be cleared and not just the registered mbeans? 

If a broker is stopped + started the caches will have previous run data in them.

```
    @Override
    protected void doStop(ServiceStopper stopper) {
        super.doStop(stopper);
        // lets remove any mbeans not yet removed
        for (Iterator<ObjectName> iter = registeredMBeans.iterator(); iter.hasNext();) {
            ObjectName name = iter.next();
            try {
                managementContext.unregisterMBean(name);
            } catch (InstanceNotFoundException e) {
                LOG.warn("The MBean {} is no longer registered with JMX", name);
            } catch (Exception e) {
                stopper.onException(this, e);
            }
        }
        registeredMBeans.clear();
    }
```

File: activemq-broker/src/main/java/org/apache/activemq/broker/jmx/ManagedRegionBroker.java [hunk 1]
```diff
@@ -215,6 +217,7 @@ public ObjectName registerSubscription(ConnectionContext context, Subscription s
                 registerSubscription(objectName, sub.getConsumerInfo(), key, view);
             }
             subscriptionMap.put(sub, objectName);
+            consumerSubscriptionMap.put(sub.getConsumerInfo(), sub);
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply activemq-broker/src/main/java/org/apache/activemq/broker/jmx/ManagedRegionBroker.java:220]: Good catch, that's possible. At least we should test if `sub` is not null.

**Comment by mattrpav** [comment]
>         registeredMBeans.clear();

We probably at least need a unit test to verify after stop that all the collections are size == 0

**Comment by lucastetreault** [comment]
Hey Matt, as discussed on slack. I'll have a look and add some tests in the next few days. Just wanted to add a comment here in case others were following along :)

**Comment by jbonofre** [comment]
@lucastetreault thanks ! Much appreciated. You can create a new PR, I will do a review. Thanks again !

**Comment by lucastetreault** [comment]
> In the doStop()-- shouldn't all the cache collections be cleared and not just the registered mbeans?
> 
> If a broker is stopped + started the caches will have previous run data in them.
> 
> ```
>     @Override
>     protected void doStop(ServiceStopper stopper) {
>         super.doStop(stopper);
>         // lets remove any mbeans not yet removed
>         for (Iterator<ObjectName> iter = registeredMBeans.iterator(); iter.hasNext();) {
>             ObjectName name = iter.next();
>             try {
>                 managementContext.unregisterMBean(name);
>             } catch (InstanceNotFoundException e) {
>                 LOG.warn("The MBean {} is no longer registered with JMX", name);
>             } catch (Exception e) {
>                 stopper.onException(this, e);
>             }
>         }
>         registeredMBeans.clear();
>     }
> ```

@mattrpav, you're correct, none of the collections are cleared when the broker stops. That seems to be the way it has been since about 2006 though. I think it might be worthwhile to clean up be can probably be done as part of a separate Jira issue. I can create a task for this if you want.

**Comment by lucastetreault** [comment]
I created [AMQ-9123](https://issues.apache.org/jira/browse/AMQ-9123) to follow up on clearing the collections when the broker is stopped. I'll make the changes soon and run all the tests to make sure things are still behaving as expected.

**Pull request merged**